hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/MarkerHandler.java

```diff
@@ -0,0 +1,297 @@
+
+import static org.apache.hudi.timeline.service.RequestHandler.jsonifyResult;
+
+public class MarkerHandler extends Handler {
+  public static final String MARKERS_FILENAME = "MARKERS";
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  private static final Logger LOG = LogManager.getLogger(MarkerHandler.class);
+  private static final long MARGIN_TIME_MS = 10L;
+
+  private final Registry metricsRegistry;
+  private final ScheduledExecutorService executorService;
+  // {markerDirPath -> all markers}
+  private final Map<String, Set<String>> allMarkersMap = new HashMap<>();
+  // {markerDirPath -> {markerFileIndex -> markers}}
+  private final Map<String, Map<Integer, StringBuilder>> fileMarkersMap = new HashMap<>();
+  private final List<CreateMarkerCompletableFuture> createMarkerFutures = new ArrayList<>();
+  private final List<Boolean> isMarkerFileInUseList;
+  private final long batchIntervalMs;
+  private final int parallelism;
+  private volatile Object createMarkerRequestlockObject = new Object();
+  private long nextBatchProcessTimeMs = 0L;
+
+  public MarkerHandler(Configuration conf, FileSystem fileSystem, FileSystemViewManager viewManager, Registry metricsRegistry,
+                       int batchNumThreads, long batchIntervalMs, int parallelism) throws IOException {
+    super(conf, fileSystem, viewManager);
+    LOG.info("*** MarkerHandler FileSystem: " + this.fileSystem.getScheme());
+    LOG.info("*** MarkerHandler Params: batchNumThreads=" + batchNumThreads + " batchIntervalMs=" + batchIntervalMs + "ms");
+    this.metricsRegistry = metricsRegistry;
+    this.batchIntervalMs = batchIntervalMs;
+    this.parallelism = parallelism;
+    this.executorService = Executors.newScheduledThreadPool(batchNumThreads);
+    List<Boolean> isMarkerFileInUseList = new ArrayList<>(batchNumThreads);
+    for (int i = 0; i < batchNumThreads; i++) {
+      isMarkerFileInUseList.add(false);
+    }
+    this.isMarkerFileInUseList = Collections.synchronizedList(isMarkerFileInUseList);
+  }
+
+  public Set<String> getAllMarkers(String markerDirPath) {
+    return allMarkersMap.getOrDefault(markerDirPath, new HashSet<>());
+  }
+
+  public Set<String> getCreateAndMergeMarkers(String markerDirPath) {
+    return allMarkersMap.getOrDefault(markerDirPath, new HashSet<>()).stream()
+        .filter(markerName -> !markerName.endsWith(IOType.APPEND.name()))
+        .collect(Collectors.toSet());
+  }
+
+  public CompletableFuture<String> createMarker(Context context, String markerDirPath, String markerName) {
+    LOG.info("Request: create marker " + markerDirPath + " " + markerName);
+    CreateMarkerCompletableFuture future = new CreateMarkerCompletableFuture(context, markerDirPath, markerName);
+    synchronized (createMarkerRequestlockObject) {
+      createMarkerFutures.add(future);
+      long currTimeMs = System.currentTimeMillis();
+      if (currTimeMs >= nextBatchProcessTimeMs - MARGIN_TIME_MS) {
```

Comment:
I will sync up w/ you offline on this code block. wanna see if we can simplify this.

Reply:
Simplified the logic to avoid the complex equation here for calculating the next batch process timestamp.